### ISSUE TITLE
Snapshot-Restored PVC with 1 Million Files Stuck Pending

### ISSUE THREAD

**Issue by tristan-le** [IBM/ibm-spectrum-scale-csi#300]
**Bug Description**
A snapshot-restored PVC with 1 Million Files is stuck in Pending state.

The original PVC is Bound:
``` bash
[root@armstrong-inf ~]# date; oc get pvc -A
Thu Aug 27 09:22:42 MST 2020
NAMESPACE             NAME                                               STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                           AGE
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1                     Bound     pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b   32Gi       RWX            210-res-4-fileset-csi-spectrum-scale   16h
210-res-4-namespace   snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1   Pending                                                                        210-res-4-fileset-csi-spectrum-scale   13h
```

The original PVC's fileset on a Worker Node contains 1 million files:
``` bash
[root@armstrong-compute4 tiny]# date; pwd; find ./ -type f | wc -l
Thu Aug 27 09:40:09 MST 2020
/ibm/ibm1/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b-data/210-res-4-fileset-blast-1-million-files-deployment1-74d5b9dnsgq_72_20-08-26-23:45:07.763284616/tiny
1040066
```

All snapshots of original PVC show 'true' for READYTOUSE state:
```bash
[root@armstrong-inf ~]# date; oc get volumesnapshot -A
Thu Aug 27 09:23:38 MST 2020
NAMESPACE             NAME                                    READYTOUSE   SOURCEPVC                        SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS         SNAPSHOTCONTENT                                    CREATIONTIME   AGE
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1-snap-1   true         210-res-4-fileset-dplmnt1-pvc1                           32Gi          210-res-4-snapclass   snapcontent-d5416290-c057-40a4-9ddc-840108fcea79   20h            13h
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1-snap-2   true         210-res-4-fileset-dplmnt1-pvc1                           32Gi          210-res-4-snapclass   snapcontent-e7e480bd-e65e-4290-9867-fce730094d97   20h            13h
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1-snap-3   true         210-res-4-fileset-dplmnt1-pvc1                           32Gi          210-res-4-snapclass   snapcontent-53aedbfa-28a1-40e1-b692-1b1d66de7947   20h            13h
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1-snap-4   true         210-res-4-fileset-dplmnt1-pvc1                           32Gi          210-res-4-snapclass   snapcontent-79dbe08b-5373-48dd-a256-1799a7ec296d   20h            13h
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1-snap-5   true         210-res-4-fileset-dplmnt1-pvc1                           32Gi          210-res-4-snapclass   snapcontent-8a833193-c127-460a-98ec-37be3ac44f99   20h            13h
```

A describe of the stuck Pending snapshot-restored PVC shows "deadline exceeded" after 116 tries and 13 hours:
``` bash
[root@armstrong-inf ~]# date; oc describe pvc -n 210-res-4-namespace snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1
Thu Aug 27 09:25:28 MST 2020
Name:          snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1
Namespace:     210-res-4-namespace
StorageClass:  210-res-4-fileset-csi-spectrum-scale
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: spectrumscale.csi.ibm.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
DataSource:
  APIGroup:  snapshot.storage.k8s.io
  Kind:      VolumeSnapshot
  Name:      210-res-4-fileset-dplmnt1-pvc1-snap-1
Mounted By:  <none>
Events:
  Type     Reason                Age                     From                                                                                                 Message
  ----     ------                ----                    ----                                                                                                 -------
  Normal   Provisioning          6m51s (x125 over 13h)   spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-0_4607666f-79b6-4787-a086-135d07d494cc  External provisioner is provisioning volume for claim "210-res-4-namespace/snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1"
  Normal   ExternalProvisioning  5m49s (x3343 over 13h)  persistentvolume-controller                                                                          waiting for a volume to be created, either by external provisioner "spectrumscale.csi.ibm.com" or manually created by system administrator
  Warning  ProvisioningFailed    4m51s (x116 over 13h)   spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-0_4607666f-79b6-4787-a086-135d07d494cc  failed to provision volume with StorageClass "210-res-4-fileset-csi-spectrum-scale": rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

**To Reproduce**
1. Create PVC with 1 Million Files
2. Take Snapshot of PVC
3. Restore PVC from Snapshot

Scripts for PVC creation, snapshots, and restore can be found here at PWD: /ibm/ibm1/real-world-tests/

**Expected behavior**
The PVC should be Bound at age 13 hours.

**Environment**
Infrastructure Node Environment:
``` bash
[root@armstrong-inf ~]# date; cat /etc/*release | grep CPE; uname -a; rpm -qi gpfs.base | grep Source; rpm -qa | grep openshift
Thu Aug 27 09:32:28 MST 2020
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
Linux armstrong-inf.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
```

OCP Cluster:
``` bash
[root@armstrong-inf ~]# date; oc get nodes --show-labels
Thu Aug 27 09:35:08 MST 2020
NAME                                                         STATUS   ROLES    AGE    VERSION           LABELS
armstrong-compute1.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute1.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute2.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute2.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute4.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute4.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute5.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute5.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute6.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute6.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute7.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute7.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute8.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute8.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-compute9.armstrong.scale-ocp.tuc.stglabs.ibm.com   Ready    worker   107d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-compute9.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/worker=,node.openshift.io/os_id=rhel,scale=true
armstrong-master1.armstrong.scale-ocp.tuc.stglabs.ibm.com    Ready    master   109d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-master1.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
armstrong-master2.armstrong.scale-ocp.tuc.stglabs.ibm.com    Ready    master   109d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-master2.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
armstrong-master3.armstrong.scale-ocp.tuc.stglabs.ibm.com    Ready    master   109d   v1.18.3+3107688   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=armstrong-master3.armstrong.scale-ocp.tuc.stglabs.ibm.com,kubernetes.io/os=linux,node-role.kubernetes.io/master=,node.openshift.io/os_id=rhcos
```

Worker Node Environments:
``` bash
[root@armstrong-compute1 ~]# date; mmdsh -f1 -N all "cat /etc/*release | grep CPE; uname -a; rpm -qi gpfs.base | grep Source; rpm -qa | grep openshift"
Thu Aug 27 09:32:27 MST 2020
armstrong-compute1-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute1-hs:  Linux armstrong-compute1.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute1-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute1-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute1-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute2-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute2-hs:  Linux armstrong-compute2.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute2-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute2-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute2-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute3-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute3-hs:  Linux armstrong-compute3.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute3-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
mmdsh: armstrong-compute3-hs remote shell process had return code 1.
armstrong-compute4-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute4-hs:  Linux armstrong-compute4.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute4-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute4-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute4-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute5-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute5-hs:  Linux armstrong-compute5.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute5-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute5-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute5-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute6-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute6-hs:  Linux armstrong-compute6.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute6-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute6-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute6-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute7-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute7-hs:  Linux armstrong-compute7.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute7-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute7-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute7-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute8-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute8-hs:  Linux armstrong-compute8.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute8-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute8-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-compute8-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute9-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-compute9-hs:  Linux armstrong-compute9.armstrong.scale-ocp.tuc.stglabs.ibm.com 3.10.0-1127.13.1.el7.x86_64 #1 SMP Fri Jun 12 14:34:17 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-compute9-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
armstrong-compute9-hs:  openshift-hyperkube-4.5.0-202007170613.p0.git.0.c785b1f.el7.x86_64
armstrong-compute9-hs:  openshift-clients-4.5.0-202007170149.p0.git.3595.faba6f6.el7.x86_64
armstrong-nsd1-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-nsd1-hs:  Linux armstrongnsd1.tuc.stglabs.ibm.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-nsd1-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
mmdsh: armstrong-nsd1-hs remote shell process had return code 1.
armstrong-nsd2-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-nsd2-hs:  Linux armstrongnsd2.tuc.stglabs.ibm.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-nsd2-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
mmdsh: armstrong-nsd2-hs remote shell process had return code 1.
armstrong-nsd3-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-nsd3-hs:  Linux armstrongnsd3.tuc.stglabs.ibm.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-nsd3-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
mmdsh: armstrong-nsd3-hs remote shell process had return code 1.
armstrong-nsd4-hs:  CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
armstrong-nsd4-hs:  Linux armstrongnsd4.tuc.stglabs.ibm.com 3.10.0-1127.el7.x86_64 #1 SMP Tue Feb 18 16:39:12 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
armstrong-nsd4-hs:  Source RPM  : gpfs.base-5.0.5-2.src.rpm
mmdsh: armstrong-nsd4-hs remote shell process had return code 1.
```

**Additional context**
A snap of the logs is resident on the system (armstrong-inf), PWD: /ibm/ibm1/CSI/CSI-2.1.0-170820/ibm-spectrum-scale-csi/tools/ibm-spectrum-scale-csi-logs_08-27-2020-09\:19\:43/

**Comment by tristan-le** [comment]
An attempt of this test on a K8s cluster resulted in the same problem, a snapshot-restored PVC stuck in Pending.

**Comment by smitaraut** [comment]
It looks like snapshot copy operation is taking longer than total wait and retry time of provisioner that is getting the pvc stuck in pending state. In the background though the operation seems to have completed. Check the contents of both source and target to verify this-

```
ls -ltr /ibm/ibm1/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b-data/
ls -ltr /ibm/ibm1/pvc-f07e5d6a-cbee-447e-bb2a-30bd9ff3bcc0/pvc-f07e5d6a-cbee-447e-bb2a-30bd9ff3bcc0-data/
```

One thing you can try is increasing the provisioner timeout value to something very large like 20min just for test-
- Delete the pending pvc
- `kubectl edit pod ibm-spectrum-scale-csi-provisioner-0`
- Change the value in "args" section from "`--timeout=2m`" to "`--timeout=20m`" and save the edit
- Perform the restore operation again

**Comment by tristan-le** [comment]
There was a fileset created for the snapshot-restored PVC, despite it being stuck in Pending state and it appears to be a replica of the original PVC.

```
[root@armstrong-inf ~]# date; oc get pvc -n 210-res-4-namespace 
Fri Sep 11 09:52:19 MST 2020
NAME                                               STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                           AGE
210-res-4-fileset-dplmnt1-pvc1                     Bound     pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b   32Gi       RWX            210-res-4-fileset-csi-spectrum-scale   15d
snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1   Pending                                                                        210-res-4-fileset-csi-spectrum-scale   15d
[root@armstrong-inf ~]# ssh armstrong-compute4
Last login: Fri Sep 11 09:46:42 2020 from arsmtrong-inf.scale-ocp.tuc.stglabs.ibm.com
[root@armstrong-compute4 ~]# date; cd /ibm/ibm1; ls -ailht | grep pvc
Fri Sep 11 09:52:25 MST 2020
  3670019 drwxrwx--x.  3 root root 4.0K Sep 11 09:47 pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b
  1572867 drwxrwx--x.  3 root root 4.0K Aug 26 19:44 pvc-f07e5d6a-cbee-447e-bb2a-30bd9ff3bcc0
[root@armstrong-compute4 ibm1]# date; cd pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b-data/210-res-4-fileset-blast-1-million-files-deployment1-74d5b9dnsgq_72_20-08-26-23\:45\:07.763284616/tiny/; find ./ -type f | wc -l
Fri Sep 11 09:52:49 MST 2020
1040066
[root@armstrong-compute4 tiny]# cd /ibm/ibm1/pvc-f07e5d6a-cbee-447e-bb2a-30bd9ff3bcc0/pvc-f07e5d6a-cbee-447e-bb2a-30bd9ff3bcc0-data/210-res-4-fileset-blast-1-million-files-deployment1-74d5b9dnsgq_72_20-08-26-23\:45\:07.763284616/tiny/; find ./ -type f | wc -l
1040066
``` bash

I deleted the pending PVC and tried to edit, but the timeout field on the pod does not seem to be a valid field to change:
```
[root@armstrong-inf ~]# date; oc edit pod -n ibm-spectrum-scale-csi-driver ibm-spectrum-scale-csi-provisioner-0
Fri Sep 11 10:09:53 MST 2020
error: pods "ibm-spectrum-scale-csi-provisioner-0" is invalid
A copy of your changes has been stored to "/tmp/oc-edit-q6g8e.yaml"
error: Edit cancelled, no valid changes were saved.
``` bash

**Comment by tristan-le** [comment]
On further suggestion from Smita, I edited timeout in the provisioner statefulset:

``` bash
[root@armstrong-inf CSI-2.1.0-170820]# date; oc edit statefulset -n ibm-spectrum-scale-csi-driver ibm-spectrum-scale-csi-provisioner
Fri Sep 11 10:47:46 MST 2020
statefulset.apps/ibm-spectrum-scale-csi-provisioner edited
```

The provisioner pod then restarted with the new timeout. After a minute, the provisioner statefulset and pod revert back to their original timeout setting. Deepak commented that this is because the operator doesn't allow for the change. This will require a dev change.

**Comment by tristan-le** [comment]
Deepak changed the timeout parameter in the operator image and provided a new build: `quay.io/deeghuge/ibm-spectrum-scale-csi-operator:timeout`. After applying the updated operator yaml, the old operator and provisioner restarted. The provisioner pod kept its new timeout settings, without reverting back:

``` bash
[root@archieprt03 fs0] 2020-09-17 11:47:49
# kubectl get pod -n ibm-spectrum-scale-csi-driver | egrep 'operator|provisioner'
ibm-spectrum-scale-csi-operator-7b659994d4-6c79k   1/1     Running   0          170m
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          168m
[root@archieprt03 fs0] 2020-09-17 11:48:15
# kubectl describe pod -n ibm-spectrum-scale-csi-driver ibm-spectrum-scale-csi-provisioner-0 | grep timeout
      --timeout=20m
```

Then I attempted to restore and the snapshot-restore PVC is again stuck in Pending:

``` bash
[root@archieprt03 fs0] 2020-09-17 11:49:53
# kubectl get volumesnapshot -n 210-r-namespace
NAME                                READYTOUSE   SOURCEPVC                    SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS     SNAPSHOTCONTENT                                    CREATIONTIME   AGE
210-r-fileset-dplmnt1-pvc1-snap-1   true         210-r-fileset-dplmnt1-pvc1                           10Mi          210-r-snapclass   snapcontent-d80da36b-a2c9-4959-8ef8-602ce500db02   8d             7d22h
210-r-fileset-dplmnt1-pvc1-snap-2   true         210-r-fileset-dplmnt1-pvc1                           10Mi          210-r-snapclass   snapcontent-0d17acbc-4d21-4f5b-890d-7c17ff46d6dc   8d             7d22h
210-r-fileset-dplmnt1-pvc1-snap-3   true         210-r-fileset-dplmnt1-pvc1                           10Mi          210-r-snapclass   snapcontent-6075b12e-c3a3-42e0-beb6-4b579531d64f   8d             7d22h
210-r-fileset-dplmnt1-pvc1-snap-4   true         210-r-fileset-dplmnt1-pvc1                           10Mi          210-r-snapclass   snapcontent-b9ec6ae3-02cb-4188-8e78-285dbad52485   8d             7d22h
210-r-fileset-dplmnt1-pvc1-snap-5   true         210-r-fileset-dplmnt1-pvc1                           10Mi          210-r-snapclass   snapcontent-7f04b498-d09f-40a4-8985-2372cb6e8df9   8d             7d22h
[root@archieprt03 fs0] 2020-09-17 11:50:03
# kubectl get pvc -n 210-r-namespace
NAME                                           STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
210-r-fileset-dplmnt1-pvc1                     Bound     pvc-1397755a-41aa-47de-a96f-7d69205f480b   10Mi       RWX            210-r-fileset-csi-spectrum-scale   7d23h
snap-210-r-fileset-dplmnt1-pvc1-snap-1-pvc-1   Pending                                                                        210-r-fileset-csi-spectrum-scale   160m
```

The following are events from describing the pending PVC:

```bash
Type     Reason                Age                    From                                                                                                 Message
  ----     ------                ----                   ----                                                                                                 -------
  Warning  ProvisioningFailed    28m                    spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-0_77c2da12-eff8-4198-b724-4b845651b1a8  failed to provision volume with StorageClass "210-r-fileset-csi-spectrum-scale": rpc error: code = DeadlineExceeded desc = context deadline exceeded
  Normal   Provisioning          4m52s (x13 over 48m)   spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-0_77c2da12-eff8-4198-b724-4b845651b1a8  External provisioner is provisioning volume for claim "210-r-namespace/snap-210-r-fileset-dplmnt1-pvc1-snap-1-pvc-1"
  Warning  ProvisioningFailed    4m52s (x12 over 28m)   spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-0_77c2da12-eff8-4198-b724-4b845651b1a8  failed to provision volume with StorageClass "210-r-fileset-csi-spectrum-scale": rpc error: code = Aborted desc = volume creation already in process : pvc-58815c66-1407-49d1-8023-d8af97ddead4
  Normal   ExternalProvisioning  3m20s (x182 over 48m)  persistentvolume-controller                                                                          waiting for a volume to be created, either by external provisioner "spectrumscale.csi.ibm.com" or manually created by system administrator
```

It appears that changing timeout may not be the fix or that timeout needs to be extended further than 20 minutes.

**Comment by tristan-le** [comment]
Trying the same fix that Deepak provided on a different cluster, Armstrong (OCP), which did have a healthy Scale instance, worked.

Applying the updated operator yaml, resulted in a restarted operator and provisioner pod, with the provisioner pod sporting the new timeout without reverting back:

``` bash
[root@armstrong-inf snapshots]# date; oc get pod -n ibm-spectrum-scale-csi-driver | egrep 'operator|provisioner'
Thu Sep 17 15:51:46 MST 2020
ibm-spectrum-scale-csi-operator-7b659994d4-64l8n   1/1     Running   0          32m
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          28m
[root@armstrong-inf snapshots]# date; oc describe pod -n ibm-spectrum-scale-csi-driver ibm-spectrum-scale-csi-provisioner-0 | grep timeout
Thu Sep 17 15:52:03 MST 2020
      --timeout=20m
```

Restoration of a PVC with a million files worked with the new timeout parameter.

``` bash
[root@armstrong-inf snapshots]# date; oc get pvc -A
Thu Sep 17 15:48:02 MST 2020
NAMESPACE             NAME                                               STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                           AGE
210-res-4-namespace   210-res-4-fileset-dplmnt1-pvc1                     Bound    pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b   32Gi       RWX            210-res-4-fileset-csi-spectrum-scale   21d
210-res-4-namespace   snap-210-res-4-fileset-dplmnt1-pvc1-snap-1-pvc-1   Bound    pvc-1899d762-1899-4810-97ae-2be988cdfa30   32Gi       RWX            210-res-4-fileset-csi-spectrum-scale   19m
[root@armstrong-inf snapshots]# ssh armstrong-compute4
Last login: Thu Sep 17 15:28:46 2020 from 9.163.6.210
[root@armstrong-compute4 ~]# cd /ibm/ibm1/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b/pvc-ce42e389-bed7-4a74-8b2f-1be4aa4d191b-data/210-res-4-fileset-blast-1-million-files-deployment1-74d5b9dnsgq_72_20-08-26-23\:45\:07.763284616/tiny/; ls
diskinfo.blt  test0  test1
[root@armstrong-compute4 tiny]# find ./ -type f | wc -l
1040066
[root@armstrong-compute4 tiny]# date; cd /ibm/ibm1/pvc-1899d762-1899-4810-97ae-2be988cdfa30/pvc-1899d762-1899-4810-97ae-2be988cdfa30-data/210-res-4-fileset-blast-1-million-files-deployment1-74d5b9dnsgq_72_20-08-26-23\:45\:07.763284616/tiny; ls
Thu Sep 17 15:48:30 MST 2020
diskinfo.blt  test0  test1
[root@armstrong-compute4 tiny]# find ./ -type f | wc -l
1040066
```

Thank you for all the help, you guys!

**Comment by Jainbrt** [comment]
@tristan-le we should keep this defect open and track until we want to have this fix in the actual GA image.

**Comment by tristan-le** [comment]
Reopened per Abhishek's request. Also, any subsequent restores did not work, only the first one did.

**Comment by deeghuge** [comment]
Hi @tristan-le, Please retry with latest dev build.

**Comment by tristan-le** [comment]
This problem persists in CSI 2.2.0 (dev branch). Using the latest dev build and the modified xcputil.sh, the problem still presents itself in attempting to restore a snapshot of a PVC with a million files.

**Comment by tristan-le** [comment]
This problem persisted in CSI 2.2.0 (dev branch, GM) on Scale 5.1.1.0 (3/29 Build). Using the latest dev build/GM on Scale 5.1.1.0 (which had a modified mmxcp), the problem still presents itself in attempting to restore a snapshot of a PVC with a million files. More Details in the story here: https://jazz07.rchland.ibm.com:21443/jazz/web/projects/GPFS#action=com.ibm.team.workitem.viewWorkItem&id=268774

**Comment by tristan-le** [comment]
This problem was resolved when tested on CSI 2.2.0 GA and Scale 5.1.1.1 (5/18 Build). Using the latest Scale 5.1.1.1 build available, the mmxcp problem was no longer present and the snapshot of a PVC with 1 million nest files (Blast) restored in 20 minutes. More details in the story here: https://jazz07.rchland.ibm.com:21443/jazz/web/projects/GPFS#action=com.ibm.team.workitem.viewWorkItem&id=268774
